### PR TITLE
vpi1-samples: add recipe

### PR DESCRIPTION
**Pull request by ichergui**
Add recipe to build and install VPI sample applications.

Signed-off-by: Ilies CHERGUI <ilies.chergui@gmail.com>

### CODE REVIEW

File: layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb [hunk 1]
```diff
@@ -0,0 +1,89 @@
+SUMMARY = "NVIDIA VPI command-line sample applications"
+HOMEPAGE = "https://developer.nvidia.com/embedded/vpi"
+LICENSE = "Proprietary"
```
Comment (madisongh) [inline layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb:3]: The code itself is BSD-3-Clause.

File: layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+LIC_FILES_CHKSUM = "file://opt/nvidia/vpi1/samples/assets/LICENSE;md5=6b5f633fc3acaabf21035790a05b1c71"
```
Comment (madisongh) [inline layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb:4]: Should also extract the BSD license text from one of the source files.

File: layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+COMPATIBLE_MACHINE = "(tegra)"
+
+inherit l4t_deb_pkgfeed cuda
+
+SRC_COMMON_DEBS = "\
+    vpi1-samples_${PV}_arm64.deb;subdir=vpi1-samples \
+"
+SRC_URI[sha256sum] = "98ea137962cb4bbfb1a778faf1cc3b19e27bbd0473aa7ab21cf9fde3a0469d9b"
+
+SRC_URI_append = " file://0001-06-klt_tracker-minor-update-in-usage.patch"
```
Comment (madisongh) [inline layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb:15]: Is this really necessary? Some of the other samples also have different command-line options than is documented.

File: layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    ${VPI_SAMPLES_PATH}/03-harris_corners \
+    ${VPI_SAMPLES_PATH}/04-rescale \
+    ${VPI_SAMPLES_PATH}/05-timing \
+    ${VPI_SAMPLES_PATH}/06-klt_tracker \
+    ${VPI_SAMPLES_PATH}/07-fft \
+    ${VPI_SAMPLES_PATH}/09-tnr \
+    ${VPI_SAMPLES_PATH}/10-perspwarp \
+    ${VPI_SAMPLES_PATH}/11-fisheye \
+    ${VPI_SAMPLES_PATH}/12-optflow_lk \
+"
```
Comment (madisongh) [inline layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb:30]: I think the whole recipe would be simpler if we add a top-level `CMakeLists.txt` file to drive the build, and then just inherit `cmake`.

File: layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    export CPPFLAGS="${CXX_EXTRA} ${CXXFLAGS} -I${CUDA_PATH}/include"
+    export LDFLAGS="-L${CUDA_PATH}/lib ${LDFLAGS}"
+
+    oldwd="$PWD"
+    for subdir in ${VPI_SAMPLES}; do
+        cd "$subdir"
+        cmake .
+        oe_runmake
+        cd "$oldwd"
+    done
```
Comment (madisongh) [inline layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb:57]: See comment above.

File: layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    chrpath -d ${D}/${VPI_SAMPLES_PATH}/bin/vpi_sample_07_fft
+    install -m 0755 ${B}/${VPI_SAMPLES_PATH}/09-tnr/vpi_sample_09_tnr ${D}/${VPI_SAMPLES_PATH}/bin
+    chrpath -d ${D}/${VPI_SAMPLES_PATH}/bin/vpi_sample_09_tnr
+    install -m 0755 ${B}/${VPI_SAMPLES_PATH}/10-perspwarp/vpi_sample_10_perspwarp ${D}/${VPI_SAMPLES_PATH}/bin
+    chrpath -d ${D}/${VPI_SAMPLES_PATH}/bin/vpi_sample_10_perspwarp
+    install -m 0755 ${B}/${VPI_SAMPLES_PATH}/11-fisheye/vpi_sample_11_fisheye ${D}/${VPI_SAMPLES_PATH}/bin
+    chrpath -d ${D}/${VPI_SAMPLES_PATH}/bin/vpi_sample_11_fisheye
+    install -m 0755 ${B}/${VPI_SAMPLES_PATH}/12-optflow_lk/vpi_sample_12_optflow_lk ${D}/${VPI_SAMPLES_PATH}/bin
+    chrpath -d ${D}/${VPI_SAMPLES_PATH}/bin/vpi_sample_12_optflow_lk
+}
```
Comment (madisongh) [inline layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb:86]: See comment above.

File: layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+RDEPENDS_${PN} += "libnvvpi1 cuda-cudart cuda-cufft tegra-libraries opencv ffmpeg"
```
Comment (madisongh) [inline layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb:88]: Most of this is unnecessary - bitbake will automatically pick up the shared library dependencies.
The ffmpeg dependency should be optional via PACKAGECONFIG.

File: layers/meta-tegrademo/conf/distro/tegrademo.conf [hunk 8]
```diff
@@ -75,4 +75,4 @@ INHERIT += "reproducible_build"
 BB_SIGNATURE_HANDLER ?= "OEEquivHash"
 BB_HASHSERVE ??= "auto"
 
-LICENSE_FLAGS_WHITELIST += "commercial_faad2"
+LICENSE_FLAGS_WHITELIST += "commercial_faad2 commercial"
```
Comment (madisongh) [inline layers/meta-tegrademo/conf/distro/tegrademo.conf:78]: This should be a separate commit. Also, I'd prefer minimizing the whitelist, restricting it to just the required packages (in this case, `ffmpeg`, and `x264` due to the dependency in the `ffmpeg` build). That means another commit with a bbappend for `opencv` that adds `libav` to its PACKAGECONFIG would also be needed, since the opencv recipe only looks for the full `commercial` whitelist flag.

**Review by madisongh — CHANGES_REQUESTED** [review]

File: layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+SRC_URI[sha256sum] = "98ea137962cb4bbfb1a778faf1cc3b19e27bbd0473aa7ab21cf9fde3a0469d9b"
+
+SRC_URI += "file://CMakeLists.txt;subdir=vpi1-samples/opt/nvidia/vpi1/samples \
+            file://0001-06-klt_tracker-minor-update-in-usage.patch"
+
+VPI_PREFIX = "/opt/nvidia/vpi1"
+EXTRA_OECMAKE = "-DCMAKE_INSTALL_PREFIX:PATH=${VPI_PREFIX}"
+
+PACKAGECONFIG ??= "${@bb.utils.contains('LICENSE_FLAGS_WHITELIST', 'commercial', 'video', bb.utils.contains('LICENSE_FLAGS_WHITELIST', 'commercial_ffmpeg', 'video', '', d), d)}"
+PACKAGECONFIG[video] = "-DBUILD_VIDEO_SAMPLES=ON,-DBUILD_VIDEO_SAMPLES=OFF,,ffmpeg"
```
Comment (madisongh) [inline layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb:21]: The `ffmpeg` addition here isn't needed, so this line should just read
```
PACKAGECONFIG[video] = "-DBUILD_VIDEO_SAMPLES=ON,-DBUILD_VIDEO_SAMPLES=OFF,,"
```

Comment (ichergui) [reply layers/meta-tegra-support/recipes-devtools/vpi/vpi1-samples_1.0.15.bb]: Done.
I think we should have only one comma, right ?
Like this:

```
PACKAGECONFIG[video] = "-DBUILD_VIDEO_SAMPLES=ON,-DBUILD_VIDEO_SAMPLES=OFF,"
```

**Review by madisongh — APPROVED** [review]
